bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyUnknownHandler.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.shelly.internal.handler;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.openhab.binding.shelly.internal.coap.ShellyCoapServer;
+import org.openhab.binding.shelly.internal.config.ShellyBindingConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ShellyUnknownHandler} implements a dummy handler for password protected or unknown devices.
+ *
+ * @author Markus Michels - Initial contribution
+ */
+@NonNullByDefault
+public class ShellyUnknownHandler extends ShellyBaseHandler {
+    private final Logger logger = LoggerFactory.getLogger(ShellyUnknownHandler.class);
+
+    public ShellyUnknownHandler(Thing thing, ShellyBindingConfiguration bindingConfig,
+            @Nullable ShellyCoapServer coapServer, String localIP, int httpPort) {
+        super(thing, bindingConfig, coapServer, localIP, httpPort);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Thing is using class {}", this.getClass());
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
```

Comment:
Why is this set to OFFLINE? Don't you need to check whether a valid configuration is set and you can reach the device?

Reply:
No, in this case the credentials are missing. It's not relevant if the thing is available, I can't connect. In this case a shellydevice is created and bound to this handler. When the user then changes the thing config the thing type is changed and a different handler gets bound.

Reply:
You do not know whether the credentials are missing or not. The provided configuration might meanwhile include them, so you will have to check this.

Reply:
- renamed class to ShellyProtectedHandler
- now calls super constructor & super.initialize (which includes changing the thing type)
- changed Handler Factory to use this class

As a result
- thing gets created using ShellyProtectedHandler
- ShellyBaseHandler.initialize() will then change the thing type so
- next time the Handler Factory uses ShellyRelayHandler or ShellyLightHandler

Just testing. I think now it how it should be.

Reply:
Excellent, thanks!

Reply:
checked in

Reply:
@kaikreuzer I think I know what happens
- as requested I removed the dependency from HandlerFactory to ThingHandler
- by moving the registerDeviceListener to HandlerFactory after the instance is created
- however, the unregister is never called, because the HandlerFactory doesn't know when the ThingHandler's dispose() was called
- the servlet will direct the next event to the HandlerFactory, which then calls onEvent() for the ThingHandler, which was already disposed

Reply:
Indeed, `unregisterDeviceListener` is never called. You will need to override `BaseThingHandlerFactory.removeHandler` and unregister the listener there.